### ISSUE TITLE
Incompatible CUDA Version with Numba in STUMPY

### ISSUE THREAD

**Issue by gerazov** [TDAmeritrade/stumpy#314]
Just wandering if it's possible to install stumpy without GPU support? I'm having issues with numba not supporting the latest CUDA. I don't plan on running the Matrix Profile on GPU, so ignoring this dependency would be a good thing ...

When I run `import matrix_profile_tools as mp` (installed via tslearn), I get the following traceback:
```python

  File ".../matrix_profile_baseline.py", line 16, in <module>
    import matrix_profile_tools as mp

  File ".../matrix_profile_tools.py", line 11, in <module>
    from tslearn.matrix_profile import MatrixProfile

  File ".../lib/python3.6/site-packages/tslearn/matrix_profile.py", line 24, in <module>
    import stumpy

  File ".../lib/python3.6/site-packages/stumpy/__init__.py", line 21, in <module>
    from .gpu_stump import gpu_stump  # noqa: F401

  File ".../lib/python3.6/site-packages/stumpy/gpu_stump.py", line 18, in <module>
    "(i8, f8[:], f8[:], i8,  f8[:], f8[:], f8[:], f8[:], f8[:],"

  File ".../lib/python3.6/site-packages/numba/cuda/decorators.py", line 136, in kernel_jit
    targetoptions=targetoptions)

  File ".../lib/python3.6/site-packages/numba/cuda/compiler.py", line 811, in __init__
    self.compile(sigs[0])

  File ".../lib/python3.6/site-packages/numba/cuda/compiler.py", line 952, in compile
    kernel.bind()

  File ".../lib/python3.6/site-packages/numba/cuda/compiler.py", line 576, in bind
    self._func.get()

  File ".../lib/python3.6/site-packages/numba/cuda/compiler.py", line 446, in get
    ptx = self.ptx.get()

  File ".../lib/python3.6/site-packages/numba/cuda/compiler.py", line 416, in get
    **self._extra_options)

  File ".../lib/python3.6/site-packages/numba/cuda/cudadrv/nvvm.py", line 548, in llvm_to_ptx
    ptx = cu.compile(**opts)

  File ".../lib/python3.6/site-packages/numba/cuda/cudadrv/nvvm.py", line 236, in compile
    self._try_error(err, 'Failed to compile\n')

  File ".../lib/python3.6/site-packages/numba/cuda/cudadrv/nvvm.py", line 254, in _try_error
    self.driver.check_error(err, "%s\n%s" % (msg, self.get_log()))

  File ".../lib/python3.6/site-packages/numba/cuda/cudadrv/nvvm.py", line 144, in check_error
    raise exc

NvvmError: Failed to compile

<unnamed> (114, 19): parse expected comma after load's type
NVVM_ERROR_COMPILATION
```

My CUDA version is 11.2, and numba supports 11.1.x. The problem is that I also have PyTorch and Tensorflow interlocked with CUDA, so rolling back is quite a hassle for something I'm not planning to use.

**Comment by seanlaw** [comment]
@gerazov Welcome to the STUMPY community and thank you for your interesting question. I can see how it can be frustrating when `numba` doesn't support your version of CUDA yet. There is a short term solution and a long term solution:

# Short Term Solution

The short term solution would be to install STUMPY from source but by making a small change in the `stumpy/__init.py` file:

First, uninstall your existing version of STUMPY. Then clone the STUMPY repository: 
```
git clone https://github.com/TDAmeritrade/stumpy.git`
cd stumpy
``` 

Next, update `line 20` of `stumpy/__init__.py` from:

https://github.com/TDAmeritrade/stumpy/blob/150d8e2af00dd199821179ca6a7d71f7adbaf726/stumpy/__init__.py#L19-L21

to

```
if cuda.is_available() and False:  # Basically, move to the `else` condition
```

And then reinstall STUMPY:

```
python -m pip install .
```

I believe that this should work. Would you mind testing this out?

# Long Term Solution

I don't have a great answer yet but it seems to me that, perhaps, we should try to catch the `NvvmError` exception, spit out a warning message, and set all of the GPU related functions as in the `else` condition:

https://github.com/TDAmeritrade/stumpy/blob/150d8e2af00dd199821179ca6a7d71f7adbaf726/stumpy/__init__.py#L27-L35

Or, in fact in your case, the driver was found but the `cudatoolkit` was the wrong version and so I can catch and warn of that.

**Comment by gerazov** [comment]
@seanlaw thanks for the quick answer! I ended up rolling back CUDA so I can't really test it, but having the check for future versions of stumpy makes sense.

I think catching the import error is the right way to go. This would warn the user about the need of rolling back if GPU support is needed :+1:

**Comment by seanlaw** [comment]
Sounds good. I'm going to leave this issue open and work on finding the best way to implement this. Thank you for bringing it to my attention!

**Comment by seanlaw** [comment]
@gerazov I am working with the wonderful people over on the `numba` team and the open issue can be tracked here:

https://github.com/numba/numba/issues/6645

0.53rc1 is at least 2-3 weeks away but they think that it would be great to include in that release and the `cuda.is_available()` check would automatically check if the CUDA version is compatible with `numba` so we wouldn't need any code changes in STUMPY once that change lands in `numba`. I will close this issue once that issue is resolved.

**Comment by gerazov** [comment]
Ok, awesome! It's a bit marginal but would be nice to have it fixed :+1:

**Comment by seanlaw** [comment]
This is completed and should be available in the next Numba release
